src/spawn_stubs.c

```diff
@@ -488,6 +502,188 @@ CAMLprim value spawn_unix(value v_env,
     }
   }
 
+  if (new_mask || old_mask)
+    pthread_sigmask(SIG_SETMASK, new_mask, old_mask);
+}
+
+#if defined(USE_POSIX_SPAWN)
+
+CAMLprim value spawn_unix(value v_env,
+                          value v_cwd,
+                          value v_prog,
+                          value v_argv,
+                          value v_stdin,
+                          value v_stdout,
+                          value v_stderr,
+                          value v_use_vfork,
+                          value v_setpgid,
+                          value v_sigprocmask)
+{
+  CAMLparam4(v_env, v_cwd, v_prog, v_argv);
+  CAMLlocal1(e_arg);
+  e_arg = Nothing;
+
+  pid_t pid;
+  int tmp_fds[3] = {0};  // invariant: initialized > 2
+
+  int e_error;
+  char *e_function = NULL;
+
+  posix_spawn_file_actions_t actions;
+  if (posix_spawn_file_actions_init(&actions))
+    caml_raise_out_of_memory();
+
+  posix_spawnattr_t attr;
+  if (posix_spawnattr_init(&attr))
+    caml_raise_out_of_memory();
+
+  struct spawn_info info;
+  init_spawn_info(&info,
+                  v_env, v_cwd, v_prog, v_argv,
+                  v_stdin, v_stdout, v_stderr, v_setpgid);
+  init_spawn_info_sigmask(&info, v_sigprocmask, NULL, NULL);
+
+  short attr_flags = POSIX_SPAWN_SETSIGMASK;
+  if (info.set_pgid) attr_flags |= POSIX_SPAWN_SETPGROUP;
+  e_error = posix_spawnattr_setflags(&attr, attr_flags);
+  if (e_error) {
+    e_function = "setpgid";
+    goto cleanup;
+  }
+
+  e_error = posix_spawnattr_setsigmask(&attr, &info.child_sigmask);
+  if (e_error) {
+    e_function = "pthread_sigmask";
+    goto cleanup;
+  }
+
+  if (info.set_pgid) {
+    e_error = posix_spawnattr_setpgroup(&attr, info.pgid);
+    if (e_error) {
+      e_function = "setpgid";
+      goto cleanup;
+    }
+  }
+
+  switch (info.cwd_kind) {
+    case INHERIT: break;
+    case PATH:
+      e_error = posix_spawn_file_actions_addchdir_np(&actions, info.cwd.path);
+      if (e_error) {
+        e_function = "chdir";
+        e_arg = Field(v_cwd, 0);
+        goto cleanup;
+      }
+      break;
+    case FD:
+      e_error = posix_spawn_file_actions_addfchdir_np(&actions, info.cwd.fd);
+      if (e_error) {
+        e_function = "fchdir";
+        e_arg = Field(v_cwd, 0);
+        goto cleanup;
+      }
+      break;
+  }
+
+  for (int fd = 0; fd < 3; fd++) {
+    int tmp_fd = tmp_fds[fd] = safe_dup(info.std_fds[fd]);
+    if (tmp_fd == -1) {
+      e_error = errno;
+      e_function = "dup";
+      goto cleanup;
+    }
+
+    e_error = posix_spawn_file_actions_adddup2(&actions, tmp_fd, fd);
+    if (e_error) {
+      e_function = "dup2";
+      goto cleanup;
+    }
+
+    e_error = posix_spawn_file_actions_addclose(&actions, tmp_fd);
+    if (e_error) {
+      e_function = "close";
+      goto cleanup;
+    }
+  }
+
+  e_error = posix_spawn(&pid, info.prog,
```

Comment:
I'm expecting that this function can block, at least for the duration of filesystem access, so it may be worth releasing the OCaml runtime lock here (so putting it inside a "blocking section" (see `caml_enter_blocking_section();`)) (after all, the primary reason we bother copying all the data structures to C land is so we can do that, I think)